### ISSUE TITLE
hugeedit doesn't recognize DATA section

### ISSUE THREAD

**Issue by chleroy** [libhugetlbfs/libhugetlbfs#30]
On 32 bits powerpc, source compiled with GCC 5.4 with binutils 2.26 and libhugetlbfs 2.20:

root@vgoip:~# ./hugeedit --data --text ./a.out
Segment 2 0x10000000 - 0x100008e4 (TEXT) default is HUGE pages
Segment 3 0x200008e4 - 0x20000aa4 () default is BASE pages

**Comment by gemorin** [comment]
You should paste the output of "objdump -h ./a.out"

**Comment by chleroy** [comment]
a.out:     file format elf32-powerpc

Sections:
Idx Name          Size      VMA       LMA       File off  Algn
  0 .interp       0000000d  10000114  10000114  00000114  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  1 .note.ABI-tag 00000020  10000124  10000124  00000124  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .hash         00000054  10000144  10000144  00000144  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  3 .dynsym       00000100  10000198  10000198  00000198  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  4 .dynstr       000000e6  10000298  10000298  00000298  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .gnu.version  00000020  1000037e  1000037e  0000037e  2**1
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  6 .gnu.version_r 00000020  100003a0  100003a0  000003a0  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  7 .rela.dyn     0000000c  100003c0  100003c0  000003c0  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  8 .rela.plt     0000003c  100003cc  100003cc  000003cc  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  9 .init         00000060  10000408  10000408  00000408  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 10 .text         00000414  10000468  10000468  00000468  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 11 .fini         00000030  1000087c  1000087c  0000087c  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 12 .rodata       00000032  100008ac  100008ac  000008ac  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 13 .eh_frame     00000004  100008e0  100008e0  000008e0  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 14 .ctors        00000008  200008e4  200008e4  000008e4  2**2
                  CONTENTS, ALLOC, LOAD, DATA
 15 .dtors        00000008  200008ec  200008ec  000008ec  2**2
                  CONTENTS, ALLOC, LOAD, DATA
 16 .jcr          00000004  200008f4  200008f4  000008f4  2**2
                  CONTENTS, ALLOC, LOAD, DATA
 17 .got2         00000034  200008f8  200008f8  000008f8  2**0
                  CONTENTS, ALLOC, LOAD, DATA
 18 .dynamic      000000d0  2000092c  2000092c  0000092c  2**2
                  CONTENTS, ALLOC, LOAD, DATA
 19 .data         00000008  200009fc  200009fc  000009fc  2**2
                  CONTENTS, ALLOC, LOAD, DATA
 20 .got          00000014  20000a04  20000a04  00000a04  2**2
                  CONTENTS, ALLOC, LOAD, CODE
 21 .plt          00000084  20000a18  20000a18  00000a18  2**2
                  ALLOC, CODE
 22 .bss          00000008  20000a9c  20000a9c  00000a18  2**2
                  ALLOC
 23 .comment      00000011  00000000  00000000  00000a18  2**0
                  CONTENTS, READONLY

**Comment by gemorin** [comment]
Can you paste "readelf -l ./a.out" as well?

**Comment by chleroy** [comment]
Elf file type is EXEC (Executable file)
Entry point 0x10000468
There are 7 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  PHDR           0x000034 0x10000034 0x10000034 0x000e0 0x000e0 R E 0x4
  INTERP         0x000114 0x10000114 0x10000114 0x0000d 0x0000d R   0x1
      [Requesting program interpreter: /lib/ld.so.1]
  LOAD           0x000000 0x10000000 0x10000000 0x008e4 0x008e4 R E 0x10000000
  LOAD           0x0008e4 0x200008e4 0x200008e4 0x00134 0x001c0 RWE 0x10000000
  DYNAMIC        0x00092c 0x2000092c 0x2000092c 0x000d0 0x000d0 RW  0x4
  NOTE           0x000124 0x10000124 0x10000124 0x00020 0x00020 R   0x4
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10

 Section to Segment mapping:
  Segment Sections...
   00     
   01     .interp 
   02     .interp .note.ABI-tag .hash .dynsym .dynstr .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .text .fini .rodata .eh_frame 
   03     .ctors .dtors .jcr .got2 .dynamic .data .got .plt .bss 
   04     .dynamic 
   05     .note.ABI-tag 
   06

**Comment by gemorin** [comment]
So the segment you're talking about is 'LOAD 0x0008e4 0x200008e4 0x200008e4 0x00134 0x001c0 RWE 0x10000000'.  It's executable that's why it's not seen as data segment (the code says a load r/w non-executable segment is a data segment).  It's because it's embedding the .got into the "data" segment.  Do you know if Is it a requirement of the powerpc ABI?

**Comment by chleroy** [comment]
As far as I know, the 32 bits powerpc ABI requires at least the .plt to be executable.

**Comment by gemorin** [comment]
.plt is executable on x86 as well.  That's not your problem.  The issue is .got.  If it is not in the ABI, then there might be a way of making a binary with RO data.  If it's part of it, you need to patch hugedit.

something like that:

```diff
--- a/hugeedit.c
+++ b/hugeedit.c
@@ -124,7 +124,11 @@ void check_remap_elf(void **elf, unsigned long *size, int wordsize)
 
 #define is_text(p) ((((p)->p_flags & (PF_R|PF_W|PF_X)) == (PF_R|PF_X)) && \
                        ((p)->p_memsz == (p)->p_filesz))
+#if defined(__powerpc__)
+#define is_data(p) (((p)->p_flags & (PF_R|PF_W|PF_X)) == (PF_R|PF_W|PF_X))
+#else
 #define is_data(p) (((p)->p_flags & (PF_R|PF_W|PF_X)) == (PF_R|PF_W))
+#endif
 
 #define update_phdrs(_BITS_)                                           \
 void update_phdrs##_BITS_(Elf##_BITS_##_Ehdr *ehdr, int remap_opts)    \

**Comment by chleroy** [comment]
Fixed by commit 6dfdac731393c3179f7365e292b3af4d28d4d84f
